### ISSUE TITLE
issue when logging %bi with "reverse-http"

### ISSUE THREAD

**Issue by bedis** [haproxy/haproxy#2797]
### Detailed Description of the Problem

I use a specifically crafted log format directive for all my traffic and it embeds a '%bi' in there:
```
log-format '"backendsrcip":"%bi" foobar'
```

for "regular" traffic, I see an expected log line, like: `"backendsrcip":"a.b.c.d" foobar`
but when my traffic passes through the reverse http backend, then the line is "truncated" to `"backendsrcip":"`. Can't see any IP address or hyphen ('-') neither the "foobar string"

### Expected Behavior

See either relevant %bi information or at least an hyphen, and the rest of the line of course.

### Steps to Reproduce the Behavior

1. setup a reverse http configuration
2. use the log-format above
3. send traffic into the "public" frontend and observe the log lines

### Do you have any idea what may have caused this?

rotten output from %bi when used in reverse http mode.

### Do you have an idea how to solve the issue?

no :(

### What is your configuration?

```haproxy
it's running in a container


global
 expose-experimental-directives
 log stdout len 4096 format raw local0 info

defaults
 log global
 mode http
 log-format '{ "clientip":"%ci", "clientport":"%cp", "datetime":"%tr", "frontend":"%f", "frontendip":"%fi", "frontendport":"%fp", "backend":"%b", "backendsrcip":"bi", "backendsrcport":"%bp", "server":"%s", "serverip":"%si", "serverport":"%sp", "serverqueue":"%sq", "tactive":"%Ta", "tsrvconnect":"%Tc", "tclientrequest":"%TR", "tserverresponse":"%Tr", "tqueue":"%Tw", "tclientidle":"%Ti", "httpstatus":"%ST", "httpmethod":"%HM", "httpuri":"%HU", "httppath":"%HPO", "httphost":"%[var(txn.host)]", "httpuseragent":"%[var(txn.useragent)]", "httpreferer":"%[var(txn.referer)]", "terminationstate":"%ts", "sslkeytype":"%[ssl_f_sig_alg]", "sslversion":"%sslv", "sslcipher":"%[ssl_fc_cipher]", "sslisresumed":"%[ssl_fc_is_resumed]", "ssl_fc_sni":"%[ssl_fc_sni]" }'
 error-log-format '{ "clientip":"%ci", "clientport":"%cp", "datetime":"%tr", "frontend":"%f", "frontendip":"%fi", "frontendport":"%fp", "fcerrorid":"%[fc_err]", "fcerrorstr":"%[fc_err_str]", "sslfcerror":"%[ssl_fc_err,hex]", "sslfcerrorstr":"%[ssl_fc_err_str]", "sslclienterror":"%[ssl_c_err]", "sslversion":"%sslv", "sslcipher":"%sslc", "sslcipherlist":"%[ssl_fc_cipherlist_str]", "sslisresumed":"%[ssl_fc_is_resumed]", "ssl_fc_sni":"%[ssl_fc_sni]"}'

```
then the rest is "standard" reverse http configuration
```


### Output of `haproxy -vv`

```plain
"official" haproxy:3.0 docker image:


HAProxy version 3.0.6-c2c0090 2024/11/07 - https://haproxy.org/
Status: long-term supported branch - will stop receiving fixes around Q2 2029.
Known bugs: http://www.haproxy.org/bugs/bugs-3.0.6.html
Running on: Linux 6.8.0-47-generic #47-Ubuntu SMP PREEMPT_DYNAMIC Fri Sep 27 21:40:26 UTC 2024 x86_64
Build options :
  TARGET  = linux-glibc
  CC      = cc
  CFLAGS  = -O2 -g -fwrapv
  OPTIONS = USE_GETADDRINFO=1 USE_OPENSSL=1 USE_LUA=1 USE_PROMEX=1 USE_PCRE2=1 USE_PCRE2_JIT=1
  DEBUG   = 

Feature list : -51DEGREES +ACCEPT4 +BACKTRACE -CLOSEFROM +CPU_AFFINITY +CRYPT_H -DEVICEATLAS +DL -ENGINE +EPOLL -EVPORTS +GETADDRINFO -KQUEUE -LIBATOMIC +LIBCRYPT +LINUX_CAP +LINUX_SPLICE +LINUX_TPROXY +LUA +MATH -MEMORY_PROFILING +NETFILTER +NS -OBSOLETE_LINKER +OPENSSL -OPENSSL_AWSLC -OPENSSL_WOLFSSL -OT -PCRE +PCRE2 +PCRE2_JIT -PCRE_JIT +POLL +PRCTL -PROCCTL +PROMEX -PTHREAD_EMULATION -QUIC -QUIC_OPENSSL_COMPAT +RT +SHM_OPEN +SLZ +SSL -STATIC_PCRE -STATIC_PCRE2 +SYSTEMD +TFO +THREAD +THREAD_DUMP +TPROXY -WURFL -ZLIB

Default settings :
  bufsize = 16384, maxrewrite = 1024, maxpollevents = 200

Built with multi-threading support (MAX_TGROUPS=16, MAX_THREADS=256, default=4).
Built with OpenSSL version : OpenSSL 3.0.15 3 Sep 2024
Running on OpenSSL version : OpenSSL 3.0.15 3 Sep 2024
OpenSSL library supports TLS extensions : yes
OpenSSL library supports SNI : yes
OpenSSL library supports : TLSv1.0 TLSv1.1 TLSv1.2 TLSv1.3
OpenSSL providers loaded : default
Built with Lua version : Lua 5.4.4
Built with the Prometheus exporter as a service
Built with network namespace support.
Built with libslz for stateless compression.
Compression algorithms supported : identity("identity"), deflate("deflate"), raw-deflate("deflate"), gzip("gzip")
Built with transparent proxy support using: IP_TRANSPARENT IPV6_TRANSPARENT IP_FREEBIND
Built with PCRE2 version : 10.42 2022-12-11
PCRE2 library supports JIT : yes
Encrypted password support via crypt(3): yes
Built with gcc compiler version 12.2.0

Available polling systems :
      epoll : pref=300,  test result OK
       poll : pref=200,  test result OK
     select : pref=150,  test result OK
Total: 3 (3 usable), will use epoll.

Available multiplexer protocols :
(protocols marked as <default> cannot be specified using 'proto' keyword)
         h2 : mode=HTTP  side=FE|BE  mux=H2    flags=HTX|HOL_RISK|NO_UPG
  <default> : mode=HTTP  side=FE|BE  mux=H1    flags=HTX
         h1 : mode=HTTP  side=FE|BE  mux=H1    flags=HTX|NO_UPG
       fcgi : mode=HTTP  side=BE     mux=FCGI  flags=HTX|HOL_RISK|NO_UPG
  <default> : mode=TCP   side=FE|BE  mux=PASS  flags=
       none : mode=TCP   side=FE|BE  mux=PASS  flags=NO_UPG

Available services : prometheus-exporter
Available filters :
	[BWLIM] bwlim-in
	[BWLIM] bwlim-out
	[CACHE] cache
	[COMP] compression
	[FCGI] fcgi-app
	[SPOE] spoe
	[TRACE] trace
```
```


### Last Outputs and Backtraces

_No response_

### Additional Information

_No response_

**Comment by Darlelet** [comment]
thanks for the report

I think  `addr_to_str()` which doesn't know how to deal with the address and does nothing.

Unfortunately `lf_ip()` doesn't expect that and calls `lf_text()` with empty string, and `lf_text()` ends up returning NULL, which aborts the whole log-format evaluation

I'll check

**Comment by Darlelet** [comment]
Bug was introduced in fcb7e4beaa

**Comment by Darlelet** [comment]
The lf_text() regression in itself was just fixed.

Regarding the fact that  `addr_to_str()` doesn't print anything for reverse http addresses, do we want to do something about that @a-denoyelle?

**Comment by Darlelet** [comment]
I also pushed an additional fix (not marked for backports to avoid breaking things in existing setups) to overcome historical behavior that caused "+M" option (responsible for "-" when no data is produced for logformat alias for instance) to be ignored in some cases. Indeed, without this last fix, while the `%bi` wouldn't truncate the log anymore, no '-' would be emitted even if "+M" option was set.

**Comment by a-denoyelle** [comment]
Yes I'll look to fix this for the reverse-http case, thanks.

**Comment by wtarreau** [comment]
I'm a bit confused by this change. First, I couldn't find the 'M' option in the doc so I don't know precisely what it promises, but for me there's a difference between an empty string and a non-existing string. A sample fetch function which returns an empty string has found the data, and it's the empty string. One that doesn't find something returns a not found. I feel like here we're considering them more or less equivalent because %bi doesn't return an address and it's seen as an empty string. Logically it should instead be a non-existing sample. I agree that both are then turned to empty string in the log format however.

The second point is that normally the "-" that is printed is for isolated empty or non-existing fields, but not for those which are part of a larger block. It's not clear to me here if concatenating variables like `%[var(txn.pfx)]%[str(blah)]%[var(txn.sfx)]` will now automatically result in "-blah-" instead of "blah" when the prefix and suffix are empty. I know we broke that a few times in the past and we need to be very careful about this, and I cannot figure the impact on this form the commit message. Because if we do that, it's definitely a breaking change on the log-fmt language and wouldn't be welcome for 3.2 (or even at all maybe).

Possibly that all of this remains OK but I couldn't figure it by myself.

**Comment by Darlelet** [comment]
You're right, the 'M' option although implemented there 723b73ad75b618467a3829e17d08290395c91f2f  and used quickly after has never been documented. But it is actually possible to override it with "+M" or "-M", thus I wonder if we should document it alongside other options such as "+Q".

Also, it is enabled by default at some places (by passing `LOG_OPT_MANDATORY` as optional argument to `parse_logformat_string()`).

> I know we broke that a few times in the past and we need to be very careful about this, 

Indeed, from what I can see in the git log, it seems that it was enabled/disabled before reaching a "stable" state (6cbbdbf3f363711b2264565763bf1b6c6a7af012, 3bfeadb3f6d0cdb924847a98669f8e908a224169)

Thus I don't know what should be done now, but it felt wrong to not evaluate the LOG_OPT_MANDATORY option consistently in `lf_text()` vs `lf_text_len()`, and e37976166b66205c4ff2530e8a2b85ae3c50524c tried to make a general rule out of it, that is, calling `lf_text()` or `lf_text_len()` with empty string will always result in "-" being produced if the option is set. 
This is the only way to trigger "-" emission, which means forcefully disabling the option with "-M" (globally or just before an alias or expression) may help to get rid of this behavior when not expected. 

Without this latter patch (e37976166b66205c4ff2530e8a2b85ae3c50524c), I do think the "M" option is error-prone, as it cannot be 100% relied upon if scripts really expect each logformat node (alias or expr) to produce something (either the data or "-" if no data or error) to help parse the log output by mapping each log value to the associated key (alias/expr) in the logformat.
Based on your example, I don't think we would be breaking something around this area though, because `%[]` doesn't leverage `lf_text()` anyway (it uses the `lf_text_len()` variant which already properly considered the "M" option). `lf_text()` is mainly used with logformat aliases.

**Comment by wtarreau** [comment]
OK, thanks for clarifying. I somewhat remember that long ago we tried hard to only produce that "-" for isolated empty fields (not surrounded by any field, only spaces or nothing), as we'd had cases where multiple consecuritve spaces in the log line wouldn't disclose which fields were written where. That was particularly visible in CLF log by then since they used to log some fields not always present. If that remains limited to aliases only, I guess we can keep it as the use case is indeed quite rare. But we need to remain super careful about that case over time!

Thanks!
